chainstate/src/detail/chainstateref.rs

```diff
@@ -422,46 +425,404 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
+        const MAX_DEC_COUNT: u8 = 18;
+        const MAX_TICKER_LEN: usize = 5;
+
+        let mut is_correct_data = true;
+
+        // Check token name
+        is_correct_data = is_correct_data
+            && (token_ticker.len() <= MAX_TICKER_LEN)
+            && (!token_ticker.is_empty())
+            && String::from_utf8_lossy(token_ticker).is_ascii();
+        //TODO: Shall we have check for unique token name?
+
+        // Check amount
+        is_correct_data = is_correct_data && amount_to_issue != &Amount::from_atoms(0);
+
+        // Check decimals
+        is_correct_data = is_correct_data && number_of_decimals < &MAX_DEC_COUNT;
+
+        // Check URI
+        is_correct_data = is_correct_data
+            && metadata_uri.len() <= MAX_URI_LEN
+            && String::from_utf8_lossy(metadata_uri).is_ascii();
+
+        match is_correct_data {
+            true => Ok(()),
+            false => Err(CheckBlockTransactionsError::TokenIssueFail(tx_id, block_id)),
+        }
+    }
+
+    fn fetch(&self, outpoint: &OutPoint) -> Result<Transaction, CheckBlockTransactionsError> {
+        let tx_index = self
+            .db_tx
+            .get_mainchain_tx_index(&outpoint.tx_id())
+            .ok()
+            .flatten()
+            .ok_or(CheckBlockTransactionsError::FetchFail)?;
+        match tx_index.position() {
+            common::chain::SpendablePosition::Transaction(tx_pos) => {
+                let tx = self
+                    .db_tx
+                    .get_mainchain_tx_by_position(tx_pos)
+                    .ok()
+                    .flatten()
+                    .ok_or(CheckBlockTransactionsError::FetchFail)?;
+
+                Ok(tx)
+            }
+            common::chain::SpendablePosition::BlockReward(_) => {
+                todo!("Block reward in tokens is not allowed")
+            }
+        }
+    }
+
+    // Get TokenId and Amount in input
+    fn map_tokens(&self, input: &common::chain::TxInput) -> Option<(TokenId, Amount)> {
+        let output_index = input.outpoint().output_index() as usize;
+        let prev_tx = self.fetch(input.outpoint()).ok()?;
+        match prev_tx.outputs().get(output_index)?.value() {
+            OutputValue::Coin(_) => None,
+            OutputValue::Asset(asset) => Some(match asset {
+                AssetData::TokenTransferV1 { token_id, amount } => (*token_id, *amount),
+                AssetData::TokenIssuanceV1 {
+                    token_ticker: _,
+                    amount_to_issue,
+                    number_of_decimals: _,
+                    metadata_uri: _,
+                } => {
+                    let token_id = token_id(&prev_tx)?;
+                    (token_id, *amount_to_issue)
+                }
+                AssetData::TokenBurnV1 {
+                    token_id: _,
+                    amount_to_burn: _,
+                } => {
+                    /* Token have burned and can't be transfered */
+                    None?
```

Comment:
Nitpick: I guess the question mark isn't necessary here?..

Reply:
I can't just remove the question mark. If I remove the question mark, then I have to add in other arms `Some`. Not sure, is it better than have question mark or not. What do you think?

Reply:
Why not?! Let's leave here `return None`